ArduCopter/mode_auto.cpp

```diff
@@ -137,36 +137,42 @@ void Copter::ModeAuto::takeoff_start(const Location& dest_loc)
 {
     _mode = Auto_TakeOff;
 
-    // convert location to class
-    Location_Class dest(dest_loc);
+    if (dest_loc.flags.local_frame) {
+        const Vector3f& curr_pos = inertial_nav.get_position();
+        // no need to check return status because terrain data is not used
+        wp_nav->set_wp_destination(Vector3f(curr_pos.x, curr_pos.y, -dest_loc.alt), false); 
```

Comment:
Because local frame is the position relative to origin, x,y,z can be any value (including zero). Unlike global frame, there are valid range for lat/lng

Reply:
Zero is actually a valid lat/lng, although ArduPilot rejects it in many places. But I don't understand what does that have to do with using the target positon...
Also, the checks I was talking about are about the Z validity: you can't go lower than you currently are and you can't takeoff to less than 100cm above home (in this case, above origin might make more sense, although I think 
this check shouldn't even exist).

Reply:
> Why doesn't local frame use the target position?

I just follow current code, because global frame use current position, too
```
Location_Class dest(dest_loc);
// set horizontal target
dest.lat = copter.current_loc.lat;
dest.lng = copter.current_loc.lng;
```